protobuf: utility methods for com.google.rpc.Status

ericgribkoff opened the pull request:
This adds utility methods for attaching a `google/rpc/status.proto` protobuf object to a gRPC `Status(Runtime)Exception` on the server-side and retrieving it on the client-side. This reuses the `Status(Runtime)Exception` metadata mechanism introduced in https://github.com/grpc/grpc-java/pull/1811. 

There was an earlier discussion in https://github.com/grpc/grpc-java/issues/681 about the appropriate mechanism to attach metadata to status responses:

The design document for mapping from gRPC `Status` to `google/rpc/status.proto` suggests that the `google/rpc/status.proto` data should be regarded as a first-class member of the gRPC `Status` API.  Since `grpc-core` cannot depend on protobuf, an alternative approach to that chosen for this PR is to introduce a `io.grpc.GoogleRpcStatusWrapper` class with member variables `int code`, `String message`, `Object[] details` that maps 1-1 with `google/rpc/status.proto `. This would allow attaching `GoogleRpcStatusWrapper` as a member variable in gRPC `Status`. 

However, we would still have to jump through a number of hoops in `grpc-core` to serialize `io.grpc.GoogleRpcStatusWrapper` to protobuf in the trailing metadata and deserialize the protobuf to `io.grpc.GoogleRpcStatusWrapper` on the client side. It seemed cleaner and much lighter weight to leverage the existing metadata trailers mechanism for `Status(Runtime)Exception`s.

@a11r @yang-g This PR does not completely meet the design requirement that the proto serialized in the metadata as `grpc-status-details-bin` *must* match the returned gRPC status code. The utility functions of the PR enforce this requirement, but there is nothing stopping the user from sidestepping these utility functions altogether and arbitrarily setting the `grpc-status-details-bin` metadata key to whatever content they want. However, since `grpc-core` will never be able to validate the content of the serialized protobuf, I couldn't see a way to enforce this requirement on the wire, regardless of the choice of the utility functions here or introducing a wrapper class for `google/rpc/status.proto`.

ericgribkoff commented:
@carl-mastrangelo any estimate for when this will get reviewed?

lukaszx0 commented:
@ericgribkoff Maybe unrelated, but I'm a bit confused why we seem to be adding _propriatary_ data structure to the project which the name seems to imply.

If it isn't meant to be _propriatary_, why call it `GoogleStatus*` ? Also seems to me like the `google/rpc/status.proto`, if intended to be generic. would be well suited for protobuf's well known types under rpc (or grpc) type.

ericgribkoff commented:
@lukaszx0 It's not intended to be proprietary. There's no need for it to have Google in the class name here, this was just a perhaps poor choice on my part to try to disambiguate it from the existing gRPC status (and other potential status types).

I've renamed it to `io.grpc.protobuf.StatusProto`, which more clearly captures it's intent - we want to have a standardized mechanism to return additional information with gRPC server responses, not necessarily just for Google APIs. The `google/rpc/status.proto` extends the existing gRPC status code and message plus a `repeated Any` details field, which can encapsulate the rich response type of any proto-based API.

lukaszx0 commented:
Sounds good, I suspected that it was just weird naming, thanks for the update - `StatusProto`is much better imo!

ejona86 commented:
We shouldn't expose rpc.Status's generated code. We need to depend on artifacts in Maven Central. I'm not sure such artifacts exist today.

ericgribkoff commented:
@ejona86 There is a maven central artifact for this. I'll update the PR. I'm assuming the same (should) apply to [`census.proto`](https://github.com/grpc/grpc-java/blob/master/services/src/main/proto/google/instrumentation/census.proto) as well?

ericgribkoff commented:
@carl-mastrangelo ping?

Re: ejona86's comment yesterday, I'm working out some issues with -Werror and depending on the maven artifact for rpc.Status, but that shouldn't require postponing review.

ericgribkoff commented:
@ejona86 I removed the status.proto and generated code, and added `com.google.api.grpc:googleapis-common-protos:0.0.3` as `libraries.google_api_protos` in `build.gradle`. This requires excluding its transitive dependencies on `io.grpc`, which is done in `protobuf/build.gradle`. I wasn't sure how to accomplish this directly in `build.gradle` to avoid the separation.

I spoke with the maintainers of the googleapis jars, and they should be able to release one without the `grpc` dependency, so handling the `exclude` elegantly will hopefully not be necessary.

carl-mastrangelo commented on protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java:
```diff
@@ -0,0 +1,161 @@
+  @Test
+  public void statusRuntimeExceptionWithMetadata() throws Exception {
+    StatusRuntimeException sre = StatusProto.toStatusRuntimeException(STATUS_PROTO, metadata);
+    verifyStatusRuntimeException(sre);
```
newline between test and verify sections

ericgribkoff replied:
Done

carl-mastrangelo commented on protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java:
```diff
@@ -0,0 +1,161 @@
+    try {
+      StatusProto.toStatusRuntimeException(STATUS_PROTO, null);
+      fail("NullPointerException expected");
+    } catch (NullPointerException expectedException) {
```
You can use this pattern if you want instead of ExpectedException, but you need to check the message of the exception to make sure its the one you expect.  Its too easy to catch the wrong NPE.

ericgribkoff replied:
Done

ericgribkoff commented on protobuf/src/main/java/io/grpc/protobuf/StatusProto.java:
```diff
@@ -0,0 +1,170 @@
+  }
+
+  private static Metadata toMetadata(com.google.rpc.Status statusProto, Metadata metadata) {
+    checkNotNull(metadata);
```
Done

ericgribkoff commented on protobuf/src/main/java/io/grpc/protobuf/StatusProto.java:
```diff
@@ -0,0 +1,170 @@
+    if (trailers != null) {
+      com.google.rpc.Status statusProto = trailers.get(STATUS_DETAILS_KEY);
+      checkArgument(
+          status.getCode().value() == statusProto.getCode(),
```
It is a wire format requirement. I think that justifies failing if seen.

ericgribkoff commented:
@ejona86 I spoke again with the google API folks. It turns out `googleapis-common-protos` has been deprecated in favor of `grpc-google-common-protos`, and this actually does include grpc-generated code. They are not planning to release an artifact containing `rpc.Status` that does not depend on `grpc`. I've updated the PR to the correct dependency with the necessary exclude statements.

@

carl-mastrangelo commented on protobuf/src/test/java/io/grpc/protobuf/StatusProtoTest.java:
```diff
@@ -57,60 +57,80 @@ public void setup() {
+    try {
+      StatusProto.toStatusRuntimeException(INVALID_STATUS_PROTO);
+      fail("IllegalArgumentException expected");
+    } catch (IllegalArgumentException expectedException) {
```
check message for code being invalid?

ericgribkoff replied:
Done, will merge on green.
